### PR TITLE
feature/search-page

### LINKED ISSUE #57: Event Meetings (and Search) Containers
Create the Event Meetings Container and Search variant

The "raw" event meetings container is basically this page:
https://councildataproject.github.io/seattle/#/events
image

But for v3 there are changes:
* Use the new Event Card Component
* Use the new Filter and Sort components
* Add the search bar that will simply re-route to the searched events page and with the same filter and sort state
* Maybe update the styling of the page selection to try to match mozilla protocol

The "searched" event meetings container is this page:
https://councildataproject.github.io/seattle/#/search?q=housing
image

With all the same details as above for v3 changes.

### PR DESCRIPTION
**Pull request by tohuynh**
### Link to Relevant Issue

This pull request resolves #57 

### Description of Changes

4 parts

#### Refactoring
- I pulled out some styled components into their own files so that they can be reused.
- useFetchData, the general data fetching react hook, now takes a function that fetches data and handles the data fetching UI status changes(loading, error, has data).

#### Search page
The search page contains a glimpse of the search results -- the top 4 most relevant event result (and later to be added the top 4 legislation results). They can click on View all N events search results to see all the event search result on a new page. 

Overall, the search page contains no filtering (see [wireframe](https://projects.invisionapp.com/share/UGW7LLV87VQ#/screens/409705102)) because I thought having filtering for events and legislations would be lead to too much functionality on one page.

I included a new page to the react-router routes `/events/search?q<query>`. Searching on the `/events` page will also take the user to `/events/search?q=<query>`.

TODO later, when the legislations index is ready, will also add `/legislations/search?q=<query>` as well.

#### UI changes
- The search bar on all pages besides the home page is much simpler. It contains a search input only (see [wireframe](https://projects.invisionapp.com/share/UGW7LLV87VQ#/screens/409705102))
- Only one filter popup can be opened at a time.
- Date range filtering is removed from the home search bar component - dates aren't directly in the events index. Users can filter by dates on the `/events/search?q=<query>` page.

#### Admin
- Installed query-string to parse `?q=<query>`
- Installed eslint plugin for linting use of react hooks

### Screenshots

`/search` page
https://user-images.githubusercontent.com/37560480/143792897-d8f0b4ff-24ee-4a36-a8f1-44870484043f.png

`/events/search?q=broadband` page
https://user-images.githubusercontent.com/37560480/143792974-87fe8208-91cc-453b-a877-3d04d10609aa.png

### CODE REVIEW

File: src/containers/SearchEventsContainer/SearchEventsContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+      <EventsFilter
+        allBodies={bodies}
+        filters={[committeeFilter, dateRangeFilter, sortFilter]}
+        sortOptions={[
+          { by: "pureRelevance", order: ORDER_DIRECTION.desc, label: "Most relevant" },
+          {
+            by: "datetimeWeightedRelevance",
+            order: ORDER_DIRECTION.desc,
+            label: "Most relevant (weighted)",
+          },
```
Comment (tohuynh) [inline src/containers/SearchEventsContainer/SearchEventsContainer.tsx:203]: What's a better label for date time weighted relevance?

**Review by JacksonMaxfield — COMMENTED** [review]
Before I even look at the code, I have some comments on the design:

First, I just want to say this PR is amazing (as are all PRs), but once again, I just want to say that as we get closer and closer to v3 it's all coming together.

1. I was originally confused as to why separate the `/search` and the `/events/search` but I get it, its for future "multi-search" with legislation. It may be useful to put a block at the bottom of the search page that simply has "coming soon" or something? right now it _feels_ odd and so as a user I am just confused as to why I need to click a button to get more results. I guess my comment here is: "we need to make it more clear that the search page _will_ do more soon and we should make the "click here for more events" link more obvious."

2. can we make the search bar the full width of the container for both the `/search` and the `/events/search` page?

3. related, because search is it's own thing, can we add a navigation tab for search along with events and people?

4. can we make the default sorting be datetime weighted for both `/search` and `/events/search`. and responding to your comment, I would call them "sort by pure relevance" "sort by relevance and recency" or something? -- hold on this

5. a part of me says we should simply get rid of the: 
image
button, per my previous comment about how this page should simply _always_ do the search for multiple types???

I _think_ that is all of my comments. A lot of them are open for discussion which leads me to probably my best "admin" comment for this review so far. We should take 20 minutes or so of meeting tomorrow to get feedback from the whole team on this and the events page in general.

**Comment by tohuynh** [comment]
If we get rid of the `search type` filtering on the home page and search page, then a need for "legislations coming soon" goes away. No legislation results will be shown until that part is implemented.

> right now it feels odd and so as a user I am just confused as to why I need to click a button to get more results.

Because the search page only provides a glimpse of the search results. This is a little out of my domain, but I think from the wireframe, it's better to keep user interaction on search results of different types separate. That is, if we allow users to filter and show more events and/or legislations on the same page, it would lead to an overload of information and buttons. There would have to be one set of filters (that are visually together) that filters both events and legislations to keep it concise. All filters would have to be somehow applicable to both events and legislations. The pagination of the events and legislation would have to be separated though -- one pagination UI for the events and another for the legislations. I feel that would look odd to me.

> I guess my comment here is: "we need to make it more clear that the search page will do more soon and we should make the "click here for more events" link more obvious."

Any suggestions to make it more obvious? We can get feedback from the rest of the team too.

>can we make the search bar the full width of the container for both the /search and the /events/search page?

I kind of like the smaller search bar more on => tablet resolution, but it's easy to change.

> related, because search is it's own thing, can we add a navigation tab for search along with events and people?

Do you mean a navigation tab for  `Search Events`? I think it's unnecessary because if a user is interested in searching only events, they would go to the events page and use the search bar there.

**Comment by JacksonMaxfield** [comment]
> If we get rid of the `search type` filtering on the home page and search page, then a need for "legislations coming soon" goes away. No legislation results will be shown until that part is implemented.
> 
> > right now it feels odd and so as a user I am just confused as to why I need to click a button to get more results.
> 
> Because the search page only provides a glimpse of the search results. This is a little out of my domain, but I think from the wireframe, it's better to keep user interaction on search results of different types separate. That is, if we allow users to filter and show more events and/or legislations on the same page, it would lead to an overload of information and buttons. There would have to be one set of filters (that are visually together) that filters both events and legislations to keep it concise. All filters would have to be somehow applicable to both events and legislations. The pagination of the events and legislation would have to be separated though -- one pagination UI for the events and another for the legislations. I feel that would look odd to me.
> 
> > I guess my comment here is: "we need to make it more clear that the search page will do more soon and we should make the "click here for more events" link more obvious."
> 
> Any suggestions to make it more obvious? We can get feedback from the rest of the team too.

These are all very good responses... A part of me wants to say, "until legislation tracking is done, the homepage search / the `/search` should simply be rerouted to `/events/search` to avoid the confusion?

I am basically trying to minimize the amount of user confusion as I was confused and I work on the project :sweat: -- We should definitely discuss this during meeting.

> > can we make the search bar the full width of the container for both the /search and the /events/search page?
> 
> I kind of like the smaller search bar more on => tablet resolution, but it's easy to change.

Ehhhh I feel like _because_ we are creating this search workflow we should emphasis it more. / Reminding the old quote from hackathon "search should be so good the user doesn't need to filter" -- right now because its not centered my eye doesn't go for it first. Maybe instead of making it the full width can we put it to the left of the filters? That way it visually feels like a filter?

> > related, because search is it's own thing, can we add a navigation tab for search along with events and people?
> 
> Do you mean a navigation tab for `Search Events`? I think it's unnecessary because if a user is interested in searching only events, they would go to the events page and use the search bar there.

Very fair.

**Comment by tohuynh** [comment]
> A part of me wants to say, "until legislation tracking is done, the homepage search / the /search should simply be rerouted to /events/search to avoid the confusion?

I'll comment out the search type filter and advanced options button on the home page and when the user searches (on the home page) the app will route to `/events/search`. 

> right now because its not centered my eye doesn't go for it first. Maybe instead of making it the full width can we put it to the left of the filters? That way it visually feels like a filter?

I rather have the full-width search bar than that. This only applies to the `/search` and `/events/search` page, but if the concern is that the smaller search bar is not prominent enough to suggest that's there is a search feature, you'd have to remember that users don't arrive at these page without a context. They would have to use the home page first to get here. I think of the smaller search bar as a reminder that users can search again without needing to go back to the home page. (I have no evidence or UX experience to support this though). I'm happy to make the search bar full-width so that it's consistent across pages.

**Comment by JacksonMaxfield** [comment]
Okay so updates from last night.

1. Seems like small search bar / how the search bar currently is stays.
2. Search from the landing page should simply reroute to `/events/search` (until legislation tracking)
3. Default search result order should be the datetime weighted and should be called "Most Relevant First" -- other options should be: "Closest Match First", "Newest First" and "Oldest First"?
4. Fixing white space / spacing of cards (can hold off for future PR imo)
5. Adding highlighting to the gram inside the chosen context span (can hold off for future PR imo)

did I miss anything?

Basically, only things from my perspective that need to change are 2 and 3 for this PR.

Also can @hawkticehurst, @BrianL3, and / or @nickmagruder review the code when you get a chance?

**Comment by tohuynh** [comment]
> did I miss anything?

Sound about right.

Those labels in 3. are good to me.

File: src/containers/SearchContainer/SearchContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,169 @@
+import React, { FC, useCallback, useMemo, useState, useRef } from "react";
```
Comment (JacksonMaxfield) [inline src/containers/SearchContainer/SearchContainer.tsx:1]: I think this entire page is basically going to sit in the repo until legislation is done. Unfortunate but makes the most sense.

**Review by JacksonMaxfield — APPROVED** [review]
Some comments with questions, mostly nitpicks.

Two general questions I could find in the code:
1. What happens if no events return from the search? Either the index hasn't been created, or just no events matched the search?
2. When I tested this locally, the search query doesn't carry over to the transcript search in the event page when you click an event. We can hold that over for a future PR or solve now, just wanted to make it clear if we should add a GitHub Issue or if it was a bug.

File: src/components/Layout/HomeSearchBar/HomeSearchBar.tsx [hunk 3]
```diff
@@ -59,48 +60,32 @@ const SearchExampleTopic = styled.p`
...
     name: SEARCH_TYPE.LEGISLATION,
-    label: "Legislation",
-    disabled: false,
-  },
-  {
-    name: SEARCH_TYPE.COUNCIL_MEMBER,
-    label: "Council Members",
+    label: "Legislations",
     disabled: false,
   },
```
[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply src/components/Layout/HomeSearchBar/HomeSearchBar.tsx:80]: Wait I just tested locally. This is removed. Just don't know where it is in the code because I am bad.

File: src/components/Layout/HomeSearchBar/HomeSearchBar.tsx [hunk 4]
```diff
@@ -2,13 +2,14 @@ import React, { ChangeEventHandler, FC, FormEventHandler, useState } from "react
 import { useHistory } from "react-router-dom";
 import styled from "@emotion/styled";
 
-import { SEARCH_TYPE } from "../../../constants/ProjectConstants";
+import { SearchEventsState } from "../../../pages/SearchEventsPage/types";
+import { SEARCH_TYPE } from "../../../pages/SearchPage/types";
 
-import { FilterPopup } from "../../Filters/FilterPopup";
+/* import { FiltersContainer } from "../../Filters/FiltersContainer";
```
[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply src/components/Layout/HomeSearchBar/HomeSearchBar.tsx:8]: Ignore like all of my comments. I see these are simply there to add back when ready.

**Comment by tohuynh** [comment]
> What happens if no events return from the search? Either the index hasn't been created, or just no events matched the search?

"No events found." gets displayed

> When I tested this locally, the search query doesn't carry over to the transcript search in the event page when you click an event. We can hold that over for a future PR or solve now, just wanted to make it clear if we should add a GitHub Issue or if it was a bug.

Yea currently missing. Add that comment to highlight the query issue.

I left the comment out code intact because I won't remember what I need to write in the future. The search type filter is visually removed though.

**Comment by JacksonMaxfield** [comment]
> > What happens if no events return from the search? Either the index hasn't been created, or just no events matched the search?
> 
> "No events found." gets displayed
> 
> > When I tested this locally, the search query doesn't carry over to the transcript search in the event page when you click an event. We can hold that over for a future PR or solve now, just wanted to make it clear if we should add a GitHub Issue or if it was a bug.
> 
> Yea currently missing. Add that comment to highlight the query issue or create a new one.
> 
> I left the comment out code intact because I won't remember what I need to write in the future. The search type filter is visually removed though.

Perfect. Done. And Perfect.

Feel free to merge whenever. :+1:
